File: fastest/fastest-run_length_encoding.hs
{-
You must apply run length encoding algorithm to the given string.

Run length encoding is a simple form of data compression, where sequences of repeating characters are replaced by a pair N, C where N is the size of the sequence and C is the repeated character.

INPUT:
Line 1: A string S to encode.

OUTPUT:
Line 1: The run length encoded version of S.

CONSTRAINTS:
0 < S < 50

EXAMPLE:
Input
abbcccdddd
Output
1a2b3c4d
-}
import Control.Arrow
import Data.List

main = interact answer

answer = concat . concatMap ((\ (a, b) -> [show a, [b]]) . (length &&& head)) . group

File: fastest/fastest-run_length_encoding.py
"""
Run length encoding

abbcccdddd -> 1a2b3c4d
"""
s = raw_input()
out = []
current = s[0]
count = 1
for c in s[1:]:
    if c == current:
        count += 1
    else:
        out.append(count)
        out.append(current)
        count = 1
        current = c
out.append(count)
out.append(current)

print ''.join(str(x) for x in out)
